File: Android/tuilivekit/livekit/src/main/java/com/trtc/uikit/livekit/component/liveListviewpager/LiveListViewPager.java
package com.trtc.uikit.livekit.component.liveListviewpager;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_DRAGGING;
import static androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_IDLE;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;

import androidx.viewpager2.widget.ViewPager2;

public class LiveListViewPager extends FrameLayout {
    private final Context             mContext;
    private       ViewPager2          mViewPager;
    private       LiveListViewAdapter mLiveListViewAdapter;

    private int   mCurrentPosition      = -1;
    private float mPositionOffset       = -1;
    private int   mWillSlideInPosition  = -1;
    private int   mWillSlideOutPosition = -1;

    public LiveListViewPager(Context context) {
        this(context, null);
    }

    public LiveListViewPager(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LiveListViewPager(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        init();
    }

    public void setAdapter(LiveListViewAdapter adapter) {
        mLiveListViewAdapter = adapter;
        mViewPager.setAdapter(adapter);
    }

    private int getCurrentItem() {
        return mViewPager.getCurrentItem();
    }

    public void enableSliding(boolean enabled) {
        mViewPager.setUserInputEnabled(enabled);
    }

    private void init() {
        mViewPager = new ViewPager2(mContext);
        RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT);
        addView(mViewPager, layoutParams);
        mViewPager.setOffscreenPageLimit(1);
        mViewPager.setOrientation(ViewPager2.ORIENTATION_VERTICAL);
        mViewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageScrollStateChanged(int state) {
                switch (state) {
                    case SCROLL_STATE_DRAGGING:
                        mWillSlideInPosition = -1;
                        mWillSlideOutPosition = -1;
                        mPositionOffset = -1;
                        break;
                    case SCROLL_STATE_IDLE:
                        onScrollEnd();
                        break;
                    default:
                        break;
                }
            }

            @Override
            public void onPageSelected(int position) {
                if (mCurrentPosition == -1) {
                    onFragmentDidSlideIn(position);
                    mCurrentPosition = position;
                    return;
                }
                if (isSliding()) {
                    onFragmentDidSlideOut(mCurrentPosition);
                    onFragmentDidSlideIn(position);
                    mCurrentPosition = position;
                }
            }

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                if (positionOffset <= 0) {
                    return;
                }
                if (isSliding()) {
                    return;
                }
                if (isSlideToPrevious(position, positionOffset)) {
                    onSlideToPrevious();
                } else {
                    onSlideToNext();
                }
                mPositionOffset = positionOffset;
            }
        });
    }

    private boolean isSlideToPrevious(int position, float positionOffset) {
        if (position < mCurrentPosition) {
            if (mPositionOffset == -1) {
                return true;
            }
            return positionOffset < mPositionOffset;
        }
        return false;
    }

    private boolean isSliding() {
        return mWillSlideInPosition != -1 || mWillSlideOutPosition != -1;
    }

    private void onSlideToNext() {
        if (mCurrentPosition >= mLiveListViewAdapter.getDataList().size() - 1) {
            return;
        }
        mWillSlideOutPosition = mCurrentPosition;
        mWillSlideInPosition = mCurrentPosition + 1;
        onFragmentWillSlideOut(mWillSlideOutPosition);
        onFragmentWillSlideIn(mWillSlideInPosition);
    }

    private void onSlideToPrevious() {
        if (mCurrentPosition <= 0) {
            return;
        }
        mWillSlideOutPosition = mCurrentPosition;
        mWillSlideInPosition = mCurrentPosition - 1;
        onFragmentWillSlideOut(mWillSlideOutPosition);
        onFragmentWillSlideIn(mWillSlideInPosition);
    }

    private void onScrollEnd() {
        if (!isSliding()) {
            return;
        }
        if (mWillSlideInPosition < 0 || mWillSlideInPosition >= mLiveListViewAdapter.getDataList().size()) {
            return;
        }
        if (mWillSlideOutPosition < 0 || mWillSlideOutPosition >= mLiveListViewAdapter.getDataList().size()) {
            return;
        }
        if (mWillSlideInPosition != getCurrentItem()) {
            onFragmentSlideInCancelled(mWillSlideInPosition);
            onFragmentSlideOutCancelled(mWillSlideOutPosition);
        }
        mWillSlideInPosition = -1;
        mWillSlideOutPosition = -1;
    }

    private void onFragmentWillSlideIn(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentWillSlideIn();
    }

    private void onFragmentDidSlideIn(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentDidSlideIn();
        if (position >= mLiveListViewAdapter.getDataList().size() - 2) {
            mLiveListViewAdapter.fetchData();
        }
    }

    private void onFragmentSlideInCancelled(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentSlideInCancelled();
    }

    private void onFragmentWillSlideOut(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentWillSlideOut();
    }

    private void onFragmentDidSlideOut(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentDidSlideOut();
    }

    private void onFragmentSlideOutCancelled(int position) {
        LiveListFragment fragment = (LiveListFragment) mLiveListViewAdapter.getFragment(position);
        if (fragment == null) {
            return;
        }
        fragment.onFragmentSlideOutCancelled();
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/features/livelist/view/liveListviewpager/LiveListViewPager.kt
package com.trtc.uikit.livekit.features.livelist.view.liveListviewpager

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import android.widget.RelativeLayout
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_DRAGGING
import androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_IDLE

class LiveListViewPager @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private lateinit var viewPager: ViewPager2
    private var liveListViewAdapter: LiveListViewPagerAdapter? = null

    private var currentPosition = -1
    private var positionOffset = -1f
    private var willSlideInPosition = -1
    private var willSlideOutPosition = -1

    init {
        initView()
    }

    fun setAdapter(adapter: LiveListViewPagerAdapter) {
        liveListViewAdapter = adapter
        viewPager.adapter = adapter
        adapter.setOnDataChangedListener {
            post {
                if (currentPosition != -1) {
                    val position = currentItem
                    onFragmentDidSlideOut(currentPosition)
                    onFragmentDidSlideIn(position)
                    currentPosition = position
                }
            }
        }
    }

    val currentItem: Int
        get() = viewPager.currentItem

    private fun initView() {
        viewPager = ViewPager2(context).apply {
            layoutParams = RelativeLayout.LayoutParams(
                LayoutParams.MATCH_PARENT,
                LayoutParams.MATCH_PARENT
            )
            offscreenPageLimit = 1
            orientation = ViewPager2.ORIENTATION_VERTICAL
            registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
                override fun onPageScrollStateChanged(state: Int) {
                    when (state) {
                        SCROLL_STATE_DRAGGING -> {
                            willSlideInPosition = -1
                            willSlideOutPosition = -1
                            positionOffset = -1f
                        }

                        SCROLL_STATE_IDLE -> onScrollEnd()

                        else -> {}
                    }
                }

                override fun onPageSelected(position: Int) {
                    if (currentPosition == -1) {
                        onFragmentDidSlideIn(position)
                        currentPosition = position
                        return
                    }
                    if (isSliding()) {
                        onFragmentDidSlideOut(currentPosition)
                        onFragmentDidSlideIn(position)
                        currentPosition = position
                    }
                }

                override fun onPageScrolled(
                    position: Int,
                    positionOffset: Float,
                    positionOffsetPixels: Int
                ) {
                    if (positionOffset <= 0) return
                    if (isSliding()) return
                    if (isSlideToPrevious(position, positionOffset)) {
                        onSlideToPrevious()
                    } else {
                        onSlideToNext()
                    }
                    this@LiveListViewPager.positionOffset = positionOffset
                }
            })
        }
        addView(viewPager)
    }

    private fun isSlideToPrevious(position: Int, positionOffset: Float): Boolean {
        if (position < currentPosition) {
            if (this.positionOffset == -1f) return true
            return positionOffset < this.positionOffset
        }
        return false
    }

    private fun isSliding(): Boolean {
        return willSlideInPosition != -1 || willSlideOutPosition != -1
    }

    private fun onSlideToNext() {
        if (currentPosition >= (liveListViewAdapter?.getDataList()?.size ?: 0) - 1) return
        willSlideOutPosition = currentPosition
        willSlideInPosition = currentPosition + 1
        onFragmentWillSlideOut(willSlideOutPosition)
        onFragmentWillSlideIn(willSlideInPosition)
    }

    private fun onSlideToPrevious() {
        if (currentPosition <= 0) return
        willSlideOutPosition = currentPosition
        willSlideInPosition = currentPosition - 1
        onFragmentWillSlideOut(willSlideOutPosition)
        onFragmentWillSlideIn(willSlideInPosition)
    }

    private fun onScrollEnd() {
        if (!isSliding()) return
        if (willSlideInPosition < 0 || willSlideInPosition >= (liveListViewAdapter?.getDataList()?.size ?: 0)) return
        if (willSlideOutPosition < 0 || willSlideOutPosition >= (liveListViewAdapter?.getDataList()?.size ?: 0)) return
        if (willSlideInPosition != currentItem) {
            onFragmentSlideInCancelled(willSlideInPosition)
            onFragmentSlideOutCancelled(willSlideOutPosition)
        }
        willSlideInPosition = -1
        willSlideOutPosition = -1
    }

    private fun onFragmentWillSlideIn(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.onFragmentWillSlideIn()
    }

    private fun onFragmentDidSlideIn(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.apply {
            onFragmentDidSlideIn()
            if (position >= (liveListViewAdapter?.getDataList()?.size ?: 0) - 2) {
                liveListViewAdapter?.fetchData()
            }
        }
    }

    private fun onFragmentSlideInCancelled(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.onFragmentSlideInCancelled()
    }

    private fun onFragmentWillSlideOut(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.onFragmentWillSlideOut()
    }

    private fun onFragmentDidSlideOut(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.onFragmentDidSlideOut()
    }

    private fun onFragmentSlideOutCancelled(position: Int) {
        (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.onFragmentSlideOutCancelled()
    }

    fun findViewByPosition(position: Int): View? {
        return (liveListViewAdapter?.getFragment(position) as? LiveListFragment)?.view
    }
}